iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1110 @@
+                }
             }
         }
 
         public override async Task SendEventAsync(Message message, CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
+
             try
             {
-                if (Logging.IsEnabled)
-                    Logging.Enter(this, cancellationToken, nameof(SendEventAsync));
+                // Topic depends on if the client is a module client or a device client
+                string baseTopicName = _moduleId == null ? _deviceToCloudMessagesTopic : _moduleToCloudMessagesTopic;
+                string TopicName = PopulateMessagePropertiesFromMessage(baseTopicName, message);
+
+                if (message.HasPayload && message.Payload.Length > MaxMessageSize)
+                {
+                    throw new InvalidOperationException($"Message size ({message.Payload.Length} bytes) is too big to process. Maximum allowed payload size is {MaxMessageSize}");
+                }
 
-                cancellationToken.ThrowIfCancellationRequested();
+                var mqttMessage = new MqttApplicationMessageBuilder()
+                    .WithTopic(TopicName)
+                    .WithPayload(message.Payload)
+                    .WithQualityOfServiceLevel(publishingQualityOfService)
+                    .Build();
 
-                EnsureValidState();
-                Debug.Assert(_channel != null);
+                MqttClientPublishResult result = await _mqttClient.PublishAsync(mqttMessage, cancellationToken).ConfigureAwait(false);
 
-                await _channel.WriteAndFlushAsync(message).ConfigureAwait(true);
+                if (result.ReasonCode != MqttClientPublishReasonCode.Success)
+                {
+                    throw new IotHubClientException($"Failed to publish the MQTT packet for message with correlation Id {message.CorrelationId} with reason code {result.ReasonCode}", true);
+                }
             }
-            finally
+            catch (Exception ex) when (ex is not IotHubClientException && ex is not InvalidOperationException && ex is not OperationCanceledException)
             {
-                if (Logging.IsEnabled)
-                    Logging.Exit(this, cancellationToken, nameof(SendEventAsync));
+                throw new IotHubClientException("Failed to send message.", true, ex);
             }
         }
 
         public override async Task SendEventAsync(IEnumerable<Message> messages, CancellationToken cancellationToken)
         {
-            foreach (Message message in messages)
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Note that this sends all messages at once and then waits for all the acknowledgements. This
+            // is the recommended pattern for sending large numbers of messages over an asynchronous
+            // protocol like MQTT
+            await Task.WhenAll(messages.Select(x => SendEventAsync(x, cancellationToken))).ConfigureAwait(false);
```

Comment:
How is error handling managed here, i.e. if one send fails.

Reply:
An AggregateException would be thrown. Is there a better way to handle something like this?

Reply:
I think a comment in the public API doc guiding users to look at the inner exception for the actual error reason will be helpful.